x-pack/plugin/ent-search/src/main/java/org/elasticsearch/xpack/application/rules/RuleQueryBuilder.java

```diff
@@ -0,0 +1,326 @@
+ * 2.0.
+ */
+package org.elasticsearch.xpack.application.rules;
+
+import org.apache.lucene.search.Query;
+import org.apache.lucene.util.SetOnce;
+import org.elasticsearch.TransportVersion;
+import org.elasticsearch.action.ActionListener;
+import org.elasticsearch.common.ParsingException;
+import org.elasticsearch.common.io.stream.StreamInput;
+import org.elasticsearch.common.io.stream.StreamOutput;
+import org.elasticsearch.core.Nullable;
+import org.elasticsearch.index.query.AbstractQueryBuilder;
+import org.elasticsearch.index.query.QueryBuilder;
+import org.elasticsearch.index.query.QueryRewriteContext;
+import org.elasticsearch.index.query.SearchExecutionContext;
+import org.elasticsearch.logging.LogManager;
+import org.elasticsearch.logging.Logger;
+import org.elasticsearch.xcontent.ConstructingObjectParser;
+import org.elasticsearch.xcontent.ParseField;
+import org.elasticsearch.xcontent.XContentBuilder;
+import org.elasticsearch.xcontent.XContentParser;
+import org.elasticsearch.xpack.application.rules.action.GetQueryRulesetAction;
+import org.elasticsearch.xpack.searchbusinessrules.PinnedQueryBuilder;
+import org.elasticsearch.xpack.searchbusinessrules.PinnedQueryBuilder.Item;
+
+import java.io.IOException;
+import java.util.ArrayList;
+import java.util.Collections;
+import java.util.LinkedHashMap;
+import java.util.List;
+import java.util.Map;
+import java.util.Objects;
+import java.util.function.Supplier;
+
+import static org.elasticsearch.xcontent.ConstructingObjectParser.constructorArg;
+import static org.elasticsearch.xcontent.ConstructingObjectParser.optionalConstructorArg;
+
+/**
+ * A query that will determine based on query context and configured query rules,
+ * whether a query should be modified based on actions specified in matching rules.
+ *
+ * This iteration will determine if a query should have pinned documents and if so,
+ * modify the query accordingly to pin those documents.
+ */
+public class RuleQueryBuilder extends AbstractQueryBuilder<RuleQueryBuilder> {
+
+    public static final String NAME = "rule_query";
+
+    private static final ParseField RULESET_IDS_FIELD = new ParseField("ruleset_ids");
+    private static final ParseField MATCH_CRITERIA_FIELD = new ParseField("match_criteria");
+    private static final ParseField ORGANIC_QUERY_FIELD = new ParseField("organic");
+
+    private final List<String> rulesetIds;
+    private final Map<String, Object> matchCriteria;
+    private final QueryBuilder organicQuery;
+
+    private final List<String> curatedIds;
+    private final Supplier<List<String>> curatedIdSupplier;
+    private final List<Item> curatedDocs;
+    private final Supplier<List<Item>> curatedDocsSupplier;
+
+    private final Logger logger = LogManager.getLogger(RuleQueryBuilder.class);
+
+    @Override
+    public TransportVersion getMinimalSupportedVersion() {
+        return TransportVersion.V_8_9_0; // TODO Update to 8.10.0
```

Comment:
The release version is decoupled from the transport version now. So it should be `TransportVersion.V_8_500_024`